sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/display/DisplayData.java

```diff
@@ -863,6 +864,14 @@ private DisplayData build() {
   }
 
   /**
+   * Create a display item for the specified key and {@link ValueProvider}.
+   */
+  public static ItemSpec<String> item(String key, ValueProvider<String> value) {
```

Comment:
How should this work for other types? When the value is available we can pull it out, but we otherwise fall back to a `String` representation.

Perhaps this API should instead take a `ValueProvider<?>` argument and convert the `value.get()` return to a String.

For other display data we allowed only a closed-set of types and expect users to format other types appropriately before passing them in. However `ValueProvider<?>` values may not be available during construction so the values cannot be converted in user code. As such, I think we must relax the closed-type restriction. `PipelineOptions` is similar (all types are allowed and there's no place for users to tell us how to format for display data), so we convert to `String`. There's logic in [`ProxyInvocationHandler`](https://github.com/apache/incubator-beam/blob/2c0d0f476ba92bf3b18bcc96f3dac47da176575d/sdks/java/core/src/main/java/org/apache/beam/sdk/options/ProxyInvocationHandler.java#L325) which can be refactored/reused here.